### PR TITLE
WIP: revisit dynamic signcolumn

### PR DESCRIPTION
**Pull request by blueyed**

### CODE REVIEW

File: src/nvim/buffer.c [hunk 1]
```diff
@@ -5084,17 +5083,16 @@ static int sign_compare(const void *a1, const void *a2)
     if (s1->id < s2->id) {
         return -1;
     }
-
     return 0;
 }
 
-int buf_signcols(buf_T *buf)
+int buf_signcols(buf_T *buf, int max_signs)
```
Comment (blueyed) [inline src/nvim/buffer.c]: better s/max_signs/max_signs_per_line/ probably.

File: src/nvim/buffer.c [hunk 2]
```diff
@@ -5112,23 +5110,33 @@ int buf_signcols(buf_T *buf)
...
-                same++;
+                cells = (int)mb_string2cells(sign_get_text(sign->typenr));
+                prev_lnum = sign->lnum;
             }
         }
 
         if (nr_signs > 0 && buf->b_signcols_max < same) {
             buf->b_signcols_max = same;
+            buf->b_signcols_cells = cells;
         }
```
Comment (blueyed) [inline src/nvim/buffer.c]: This loop should be made clearer probably.

File: src/nvim/edit.c [hunk 3]
```diff
@@ -5899,7 +5899,7 @@ comp_textwidth (
       textwidth -= 1;
     }
     textwidth -= curwin->w_p_fdc;
-    textwidth -= win_signcol_count(curwin);
+    textwidth -= win_signcol_width(curwin);
```
Comment (blueyed) [inline src/nvim/edit.c]: Should/could address the buffer value directly probably for performance reasons.

File: src/nvim/ex_cmds.c [hunk 4]
```diff
@@ -5721,14 +5720,7 @@ void ex_sign(exarg_T *eap)
...
             xfree(sp->sn_text);
-            // Allocate one byte more if we need to pad up
-            // with a space.
-            len = (int)(p - arg + ((cells == 1) ? 1 : 0));
-            sp->sn_text = vim_strnsave(arg, len);
-
-            if (cells == 1) {
-              STRCPY(sp->sn_text + len - 1, " ");
-            }
+            sp->sn_text = vim_strnsave(arg, p - arg);
```
Comment (blueyed) [inline src/nvim/ex_cmds.c]: This is in a separate commit - not strictly necessary, but allows for having signs only take one cell after all.

File: src/nvim/option.c [hunk 5]
```diff
@@ -3035,6 +3035,8 @@ did_set_string_option (
     // 'signcolumn'
     if (check_opt_strings(*varp, p_scl_values, false) != OK) {
       errmsg = e_invarg;
+    } else {
+      curwin->w_buffer->b_signcols_max = -1;
```
Comment (blueyed) [inline src/nvim/option.c]: Invalidates the cache - could need a comment maybe.  But it is done like this in many places already.

File: src/nvim/option.c [hunk 6]
```diff
@@ -7094,34 +7096,46 @@ int csh_like_shell(void)
...
-  return MIN(maximum, needed_signcols);
+/// Return the number of requested sign column cells, based on current
+/// buffer signs and user configuration.
+int win_signcol_cells(win_T *wp)
+{
+  int cols = win_signcol_count(wp);
+  if (cols) {
+    return MAX(cols, wp->w_buffer->b_signcols_cells);
+  }
+  return 0;
```
Comment (blueyed) [inline src/nvim/option.c]: This was quick and dirty.
Any idea how to improve this?  (not having two functions)

File: src/nvim/screen.c [hunk 7]
```diff
@@ -2709,32 +2709,26 @@ win_line (
...
-                          c_final = NUL;
-                          // symbol(s) bytes + (filling spaces) (one byte each)
-                          n_extra = symbol_blen +
-                            (win_signcol_width(wp) - mb_string2cells(p_extra));
-                          memset(extra, ' ', sizeof(extra));
-                          STRNCPY(extra, p_extra, STRLEN(p_extra));
-                          p_extra = extra;
-                          p_extra[n_extra] = NUL;
+                          n_extra = (int)STRLEN(p_extra);
+                          sign_cells -= mb_string2cells(p_extra);
```
Comment (blueyed) [inline src/nvim/screen.c]: The previous use of `extra` was not clear to me.  Using just `p_extra` and `n_extra` seems to be good, no?
/cc @teto (via 0848add48).

File: src/nvim/buffer.c (hunk 1 above)
Comment (justinmk) [reply src/nvim/buffer.c]: Don't need the long name. Sign width is by definition per-line.

File: src/nvim/option.c (hunk 5 above)
Comment (justinmk) [reply src/nvim/option.c]: probably need `SIGNCOLS_UNKNOWN` in `sign_defs.h` by now.

File: src/nvim/screen.c [hunk 8]
```diff
@@ -2661,7 +2660,7 @@ win_line (
     extra_check = true;
   }
 
-  int sign_idx = 0;
+  int sign_idx = 0, sign_cells = win_signcol_width(wp);
```
Comment (justinmk) [inline src/nvim/screen.c]: we generally do not declare multiple symbols on one line.  Specially for non-constant initializer.

File: src/nvim/buffer_defs.h [hunk 9]
```diff
@@ -781,6 +781,7 @@ struct file_buffer {
   signlist_T *b_signlist;       // list of signs to draw
   int b_signcols_max;           // cached maximum number of sign columns
   int b_signcols;               // last calculated number of sign columns
+  int b_signcols_cells;         // number of cells used for sign columns
```
Comment (justinmk) [inline src/nvim/buffer_defs.h]: why do we need another field (and thus all the associated housekeeping)? why can't this replace `b_signcols`?

**Comment by justinmk** [comment]
cc @da-x

**Comment by blueyed** [comment]
So one test is failing, but I cannot reproduce it (likely because the real UI does not allow to move the cursor there?!):
```
[  ERROR   ] test/functional/ui/sign_spec.lua @ 118: Signs :sign place multiple signs #9295
test/functional/ui/screen.lua:560: Row 4 did not match.
Expected:
  |{1:>>}XX{6:  1 }a                                            |
  |XX{1:>>}{6:  2 }b                                            |
  |{1:>>}WW{6:  3 }c                                            |
  |*{2:    }{6:  4 }^                                             |
  …
Actual:
  |{1:>>}XX{6:  1 }a                                            |
  |XX{1:>>}{6:  2 }b                                            |
  |{1:>>}WW{6:  3 }c                                            |
  |*{2:    }{6:  ^4 }                                             |
```

File: src/nvim/buffer_defs.h (hunk 9 above)
Comment (blueyed) [reply src/nvim/buffer_defs.h]: count != cols, max/b_signcols holds the count.
https://github.com/blueyed/neovim/blob/6d5f8c7addcc4b276a436531e75cb3b620fe513e/src/nvim/buffer.c#L5163-L5167
But it certainly should at least use "cols" (effective).
I might be missing something here though.

**Comment by blueyed** [comment]
Closing for now.

**Pull request closed without merging**